### ISSUE TITLE
[BUG/ISSUE] GCHP v14.1.0 Fortran runtime error: Recursive call to nonrecursive procedure

### ISSUE THREAD

**Issue by Twize** [geoschem/gchp#294]
### What institution are you from?
Hi, I am Tyler Wizenberg from the University of Toronto, working with Dylan Jones and Kimberly Strong. 

### Description of the problem
I am trying to get GCHP v14.1.0 up and running on Scinet Niagara, and I was able to compile the GCHP executable without any issues or errors. I am using the same software environment which had previously worked for me compiling and running GCHP v13.3.2. However, at runtime, I encounter the following error: 

`Error termination. Backtrace:
At line 36 of file /scratch/d/dylan/tylerw/GCHP_v14.1.0_testenv/src/yaFyaml/src/Nodes/Mapping.F90
Fortran runtime error: Recursive call to nonrecursive procedure '__copy_fy_mapping_Omap_i_s_node'`

The full error traceback is included in the gchp.log file below, but based on the first line of the full traceback, it appears the error is from the /src/yaFyaml/src/Nodes/Mapping.F90 program. I had a look in that .F90 file but I don't even see a reference to the procedure mentioned in the traceback. 

I also tried using a newer software stack with GCC 10.3.0 + OpenMPI 4.1.1, but for some reason it would just hang at runtime and never actually execute the `mpirun -np ${nCores} ./gchp` line from my runscript (I think the issue was with the newer OpenMPI), hence why I resorted to my older software stack which I knew worked. 

Many thanks in advance!

### GEOS-Chem version
GCHP v14.1.0

### Log files
- Build log (if applicable):[ecbuild.log](https://github.com/geoschem/GCHP/files/10871713/ecbuild.log)
- Cmake Lists: [CMakeLists.txt](https://github.com/geoschem/GCHP/files/10871718/CMakeLists.txt)
- CmakeCache: [CMakeCache.txt](https://github.com/geoschem/GCHP/files/10871711/CMakeCache.txt)
- Run log (if applicable): [gchp.log](https://github.com/geoschem/GCHP/files/10871697/gchp.log)

### Software versions
- Compilers (Intel or GNU, and version): GCC 9.2.0
- Cmake: v3.16.1
- NetCDF version: Netcdf-fortran 4.6.0
- MPI version: OpenMPI 4.0.3
- ESMF version: v8.3.1

**Comment by yantosca** [comment]
Thanks for writing @Twize.  I believe that you need at least GFortran 9.3.0 for GCHP now.  This is because the YAML parser (yafYaml) uses some object-oriented features in the latest Fortran standard, which are not in gfortran versions prior to 9.3.0.

We ran into this issue too..

Know that the MAPL libraries often chase the latest compiler versions as they rely on new whiz-bang Fortran features.

**Comment by yantosca** [comment]
Also we have been able to use gcc/gfortran 10.2.0 and OpenMPI 4.1.0 in our testing/benchmarking configuration.  Not sure why it is hanging in your case. MPI builds are often sensitive on things like dependencies and the OS version.  Maybe that's a question for your sysadmin.

Tagging @lizziel @kilicomu

**Comment by lizziel** [comment]
> I also tried using a newer software stack with GCC 10.3.0 + OpenMPI 4.1.1, but for some reason it would just hang at runtime and never actually execute the mpirun -np ${nCores} ./gchp line from my runscript (I think the issue was with the newer OpenMPI), hence why I resorted to my older software stack which I knew worked.

I have also run into this problem recently when I did not have a restart file for the date I am running. It's something that needs to be fixed in our sample run scripts. Try adding `set -xe` to your run script and running again. This will print information about where it gets in the run script and show where it is stopping. Also check your `Restarts` directory to see if you have a restart file for the date you are running at.

**Comment by lizziel** [comment]
Also, sorry about the confusion about minimum library versions. I have updated the minimum compiler and ESMF versions in ReadTheDocs for 14.1.1 which will be released soon.

**Comment by Twize** [comment]
Hi @yantosca and @lizziel, thanks for the quick replies!

> Thanks for writing @Twize. I believe that you need at least GFortran 9.3.0 for GCHP now. This is because the YAML parser (yafYaml) uses some object-oriented features in the latest Fortran standard, which are not in gfortran versions prior to 9.3.0.
> 
> We ran into this issue too..
> 
> Know that the MAPL libraries often chase the latest compiler versions as they rely on new whiz-bang Fortran features.

Thanks for letting me know. This is the main reason I was originally trying a newer software stack with GCC 10.3.0 + OpenMPI 4.1.1, @lizziel previously suggested to me that I should move to gcc 10+ if possible. I am not quite sure why its hanging, but I reached out to the Scinet support staff so hopefully they will be able to help me diagnose that particular issue.  If I can't get the environment with GCC 10.3.0 to work, I can maybe try one with GCC 9.3.0 to at least see if it will run without the error I mentioned above. 

> > I also tried using a newer software stack with GCC 10.3.0 + OpenMPI 4.1.1, but for some reason it would just hang at runtime and never actually execute the mpirun -np ${nCores} ./gchp line from my runscript (I think the issue was with the newer OpenMPI), hence why I resorted to my older software stack which I knew worked.
> 
> I have also run into this problem recently when I did not have a restart file for the date I am running. It's something that needs to be fixed in our sample run scripts. Try adding `set -xe` to your run script and running again. This will print information about where it gets in the run script and show where it is stopping. Also check your `Restarts` directory to see if you have a restart file for the date you are running at.

I do have a restart for the date (I renamed one of the initial C48 restart files from my v13.3.2 simulations) to match the date, and then ran ./setRestartLink.sh in my runscript which seemed to work fine. However, I don't think its the same issue you describe @lizziel . When using my GCC 10.3.0 environment it seems to hang before it even properly executes the mpirun command (when its distributing the tasks in slurm to the various nodes), and the ./gchp executable is never actually executed. I turned on all of the mpirun verbose flags and I've attached the logfile just below (theres no error message or anything, it just never progresses past that final line):

gchp.log (GCC 10.3.0): [gchp.log](https://github.com/geoschem/GCHP/files/10873932/gchp.log)

For comparison, here is a logfile with all of the mpirun verbose flags on when I use my GCC 9.2.0 environment:

gchp.log (GCC 9.2.0): [gchp.log](https://github.com/geoschem/GCHP/files/10873952/gchp.log)

**Comment by lizziel** [comment]
Yes, this does look like some sort of OpenMPI issue. We haven't had a report of that before. Definitely keep us posted on whether your sysadmins find a fix.

**Comment by kilicomu** [comment]
Hey @Twize - just to double-check...you're launching GCHP by calling `mpirun` directly? As opposed to using `srun --mpi`?

Whichever one you're running, and if your system supports e.g. `srun --mpi="pmix"`, try the other one. You can check if your system supports MPI through `srun` by running:

```
$ srun --mpi="list"
```

at the shell on your system.

**Comment by kilicomu** [comment]
I'd also try to isolate MPI as the cause by building a toy MPI programming and running that. You can find one here:

https://mpitutorial.com/tutorials/mpi-hello-world/

including some details about how to build it, if you're not familiar with that.

**Comment by Twize** [comment]
Hi @kilicomu, when I was running GCHP v13.3.2 earlier on, I was running it by calling mpirun directly and it worked fine. I think when I call mpirun, it actually executes srun under the hood (I can see this in the gchp.log files I included above): 

`[nia0004.scinet.local:143418] [[49797,0],0] plm:slurm: final top-level argv:
	srun --ntasks-per-node=1 --kill-on-bad-exit --nodes=2 --nodelist=nia0309,nia1517 --ntasks=2 orted -mca ess "slurm" -mca ess_base_jobid "3263496192" -mca ess_base_vpid "1" -mca ess_base_num_procs "3" -mca orte_node_regex "nia[4:4,309,1517]@0(3)" -mca orte_hnp_uri "3263496192.0;tcp://172.16.1.4,10.20.1.4:53811" --mca plm_base_verbose "100" --mca ras_base_verbose "100" --mca rss_base_verbose "100" --mca rmaps_base_verbose "100"
`

However, I also tried explicitly using srun (instead of mpirun) in my runscript to execute the gchp executable, but it still hangs in the same way. 

The MPI hello world program is a good suggestion, I will give this a try to see if I can isolate OpenMPI as the cause of the hangup.

**Comment by Twize** [comment]
Hi @lizziel @yantosca @kilicomu, I put together a new software environment using GCC 9.4.0 + OpenMPI 4.1.1 and it seems to compile and run the model now without the originally mentioned error, so it was likely that I was previously using GCC 9.2.0 when 9.3.0+ was required. I still haven't been able to figure out why the software environment I was using with GCC 10.3.0 + OpenMPI 4.1.1 hangs at runtime but at least it seems I'm able to run v14.1 of the model now.

However, when I run the model now, I get the following error while its initializing:
`pe=00018 FAIL at line=02176    ExtDataGridCompMod.F90                   <From ./HcoDir/AEIC2019/v2022-03/2019_monmean/AEIC_monmean_2019%m2.0.5x0.625.36L.nc could not find time no extrapolation>`

I searched around for a similar issue from other users but could not find anything. I didn't modify the HEMCO_Config.rc settings relating to the AEIC emissions, so not sure what is the cause (maybe the input file?). Have you seen this error before? I can create a new issue if needed. I have attached my slurm run log and gchp.log below.

Slurm run log: [slurm_output_8966779.txt](https://github.com/geoschem/GCHP/files/10884130/slurm_output_8966779.txt)
gchp log: [gchp.log](https://github.com/geoschem/GCHP/files/10884131/gchp.log)

**Comment by kilicomu** [comment]
Hi @Twize - glad you've got a configuration that allows you to at least get past the error you were seeing. Did you figure out whether or not a simple MPI program also hangs with the GCC 10.3.0 stack?

A separate issue for the new problem would probably be best, just helps with the searchability of issues for other users!

**Comment by Twize** [comment]
Hi @kilicomu, I didn't get a chance to try running the toy MPI program just yet. I have a feeling that something may have gone wrong when I was setting up that software stack with Spack, and I found that there were a bunch of extra flags/configuration options added to that external OpenMPI entry in packages.yaml used by Spack from when I ran `spack external find`. I might try re-making that software stack from scratch today to see if that fixes it. 

As for the issue I just mentioned, I think I found the cause. It looks like between GCHP v13.3.2 (which I was using before) and v14.1.0, the AEIC emissions were updated from v2015 to v2019, so I don't have the AEIC2019 files on our server, and there were some extra fields added for this in HEMCO_Config.rc. I think if I pull those files down it should fix it.

**Comment by Twize** [comment]
I'm going to go ahead and close this issue since it seems it was caused by using GCC <9.3.0 to compile GCHP v14.1.0. Thanks for your help @lizziel @yantosca @kilicomu!

**Comment by Twize** [comment]
Hi @kilicomu quick update. I tried setting up the GCC 10.3.0 + OpenMPI 4.1.1 software stack completely from scratch, but it still hangs at the mpirun line when trying to execute the runscript. I also tried your suggestion with the toy OpenMPI hello world program, but that hangs as well, so I think the OpenMPI 4.1.1 compiled with GCC 10.3.0 on Scinet Niagara must not be configured correctly or something. I will let the sysadmin know.
